### PR TITLE
Set up D-Bus connection with Arctica Greeter

### PR DESCRIPTION
**Pull request by tari01**
depends https://github.com/ArcticaProject/arctica-greeter/pull/51

### CODE REVIEW

File: src/keyboard-x11.c [hunk 1]
```diff
@@ -380,19 +380,18 @@ void keyboard_SetLayout(Keyboard *pKeyboard, gint nLayout)
...
-        // TODO
         gchar *sCommand;
         gchar *sId = g_slist_nth_data(pKeyboard->pPrivate->lLayoutRec, nLayout);
         gchar **lParamas = g_strsplit(sId, "+", -1);
         guint nParams = g_strv_length(lParamas);
 
         if (nParams == 1)
         {
-            sCommand = g_strdup_printf("setxkbmap -layout %s", lParamas[0]);
+            sCommand = g_strdup_printf("setxkbmap -display :0 -layout %s", lParamas[0]);
```
Comment (sunweaver) [inline src/keyboard-x11.c:390]: Hardcoding :0 as display number will work for 95% of all cases, but LightDM can also fireup a DISPLAY :1. Also on multihead systems :0 will not always work.

File: src/keyboard-x11.c [hunk 2]
```diff
@@ -380,19 +380,18 @@ void keyboard_SetLayout(Keyboard *pKeyboard, gint nLayout)
...
         }
         else
         {
-            sCommand = g_strdup_printf("setxkbmap -layout %s -variant %s", lParamas[0], lParamas[1]);
+            sCommand = g_strdup_printf("setxkbmap -display :0 -layout %s -variant %s", lParamas[0], lParamas[1]);
```
Comment (sunweaver) [inline src/keyboard-x11.c:394]: same here

**Review by sunweaver — COMMENTED** [review]
We might need to become more generic here and not assume that LightDM only can do a :0 DISPLAY.

File: src/keyboard-x11.c [hunk 3]
```diff
@@ -380,19 +380,29 @@ void keyboard_SetLayout(Keyboard *pKeyboard, gint nLayout)
     }
     else
     {
-        // TODO
+        gchar *sDisplay = ":0";
+        Display *pDisplay = XOpenDisplay (sDisplay);
+
+        if (pDisplay == NULL)
+        {
+            sDisplay = ":1";
```
[the remark this answers is not in the document]
Comment (tari01) [reply src/keyboard-x11.c:388]: I will look into this, however, I think we might eventually need to offload certain tasks to Arctica, since some applications/commands simply cannot be run properly from within indicators.

Having done the framework for the a11y indicator, it seems to me that the most elegant way would be ask Actica over D-Bus to change the layout for us, start Orca/Onboard, change the theme, etc.

I'll revisit my suggestion after I see what `dbus-update-activation-environment` can do for us.

**Comment by sunweaver** [comment]
@tari01 to be more clear about the keyboard indicator's idea when running in the greeter... 

Valuable information: the greeter (arctica-greeter) has two modes: login prompt only (without user list to scroll through) and login via selecting one's user from a presented user list.

case 1 (login prompt box only):

  * greeter starts with system locale and keyboard settings
  * user can use the keyboard indicator to adjust the keyboard layout
  * once username (and password) have been successfully given, the keyboard layout should be propagated into the user session (via AccountsService)
  * that means: greeter or greeter's keyboard indicator stores the keyboard layout in actservice and user session's keyboard indicator picks it up again and applies the keyboard layout to the user session

case 2 (picking one's user account from scrollable user list):

  * whenever a user gets selected from the user list, the greeter or greeter's keyboard indicator queries the language layout settings of that user and applies it
  * users that don't have a language stored yet, get presented with the systems default kbd layout
  * if the keyboard indicator is used to adjust the layout, it should be applied to the running greeter session
  * if login as that user is successfull, the new kbd layout should stored via AccountsService under the logging in username
  * in the user session then the keyboard indicator should pick up the kbd layout from AccountsService and re-apply it to the user session

Basically, we'll need a pretty similar mechanism some time later for language selection and hand-over to the user session.

**Comment by sunweaver** [comment]
> Basically, we'll need a pretty similar mechanism some time later for language selection and hand-over to the user session.

"language" as in locale setting for the user session (the language in what dialogs and menus etc. are displayed)

**Comment by tari01** [comment]
Thanks for the breakdown. Here' s the situation:

> case 1 (login prompt box only):
> 
>     * greeter starts with system locale and keyboard settings
Already happening.
>     * user can use the keyboard indicator to adjust the keyboard layout
Happening, but hakish.
>     * once username (and password) have been successfully given, the keyboard layout should be propagated into the user session (via AccountsService)
Already happening.
>     * that means: greeter or greeter's keyboard indicator stores the keyboard layout in actservice and user session's keyboard indicator picks it up again and applies the keyboard layout to the user session
Already happening.
> 
> case 2 (picking one's user account from scrollable user list):
> 
>     * whenever a user gets selected from the user list, the greeter or greeter's keyboard indicator queries the language layout settings of that user and applies it
Yep. This is where I'll need to create two-way D-Bus chat between Artica and the Indicator
>     * users that don't have a language stored yet, get presented with the systems default kbd layout
Already happening.
>     * if the keyboard indicator is used to adjust the layout, it should be applied to the running greeter session
Happening, but hakish.
>     * if login as that user is successfull, the new kbd layout should stored via AccountsService under the logging in username
Already happening.
>     * in the user session then the keyboard indicator should pick up the kbd layout from AccountsService and re-apply it to the user session
Already happening.

Conclusion: drop the `dbus-update-activation-environment` approach and concetrate on setting up D-Bus for Arctica <-> Indicator to be able to query, get signals and tell Arctica to do things for us (a11y will also like this). Mike agrees?

**Comment by sunweaver** [comment]
> Mike agrees?

Ok then. Let's go this route. Let's make it so generic, that also desktop envs could possibly use it later on.

File: src/keyboard-x11.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (tari01) [reply src/keyboard-x11.c]: Fixed.

**Comment by tari01** [comment]
@sunweaver: Review ping.

**Review by sunweaver — APPROVED** [review]
Will merge this PR without testing it. I will test this once packages have landed on jenkins.artica-project.org.
